python/lib/core/dmod/core/common/reader.py

```diff
@@ -5,3 +5,12 @@
 class Reader(Protocol):
     def read(self, size: int = -1, /) -> bytes:
         """EOF if empty b''."""
+
+
+class RepeatableReader(Reader):
```

Comment:
Would the convention impose on us to also implement [`seekable()`](https://docs.python.org/3/library/io.html#io.IOBase.seek) (and, from that, `tell()` and `truncate()`)?

Reply:
Generally, if `seek` is an available method, it follows that the methods you listed are also available. But, I don't think that is a strict requirement. I try to keep `Protocol` interfaces as thin and narrowly scoped as necessary. So, in this case I would not include those methods.

Reply:
I don't want to get too carried away with this, but I find the underlying design philosophy question genuinely interesting:  if the strength of adhering to convention is encoded expectations, (when) is it better avoid the convention if expectations aren't going to be fulfilled?

We can potentially continue the discussion, here or offline, but for the sake of practicality I'll plan to make the changes as you've suggested.

Reply:
Yeah! I think it's a really interesting question too! 

> if the strength of adhering to convention is encoded expectations, (when) is it better avoid the convention if expectations aren't going to be fulfilled?

I would argue that a function or method that supports a polymorphic argument should constrain the requirements of that argument to only what the function or method _needs_. So in this context, an interface (python Protocol) is simply a contract that some type `T` supports a given set of methods. It divorces the requirement of a `T` to be a certain type or subtype, but instead focuses only on the behaviors of the `T`. This enables constructing expectations (methods) for only what is _required_ and often reduces coupling. 

With this in mind, we can consider a convention as a _composition_ of expectations (methods) rather than a single set of expectations (traditional abc). So, the convention is not lost, we simply _only_ specify _our_ minimal set of requirements.

Reply:
I largely agree, but a question follows from this on the strength of transitive expectations/requirements:  does `seek()` imply `seekable()`, or merely suggest it?  

As you point out, we could answer that by noting that the utilizing method doesn't strictly need to check whether the object is seek-able.  It _does_ need something that `seekable()` would provide (assurance that the data object can be read more than once), but not all of it, and what the method would need of it can be provided in a different way.   So, we don't really need the object to include `seekable()`.

The problem is that we could apply the same rationale with `seek()`.  The method does not need an object with full seek capabilities:  only an object that can do something that could be accomplished using `seek()`.  And we could provide that functionality to the object in a different, more focused way.  So we don't _really_ need `seek()` either. 

In other words, we would not be adopting that approach because it was the strictest possible constraint fulfillment, but rather because it's familiar:   this will probably let people (including us in the future) work with the object more quickly and easily.  But that fundamentally relies on expectations and assumptions for how things work.   And in that context, I'm not sure we can get a firm answer on the implication versus suggestion question.  

So the problem here to me is to approximating the balance point between convenience at the cost of risking reasonable-but-incorrect assumptions, versus avoiding ambiguity at the cost of familiarity.  I'm still not sure where that is with this.  Your option does feel more "Pythonic," though, so absent a more definitive mechanism, we can split the tie that way.